### DISCUSSION TITLE
Can't add proper scene detection to the VSMLRT?

### DISCUSSION THREAD

**Discussion by VisionHDR** [AmusementClub/vs-mlrt#122] [General]
Hello, im trying to understand why vsmlrt does not have sc = True parameter?

How do i add proper scene detection then ? I tried: 

script_content([
                    "# Calculate padding to ensure 32-pixel alignment",
                    "ph = (clip.height + 31) // 32 * 32 - clip.height",
                    "pw = (clip.width + 31) // 32 * 32 - clip.width",
                    "",
                    "# Add padding (clip already in RGB format)",
                    "clip = clip.std.AddBorders(right=pw, bottom=ph)",
                    "",
                    "# Setting color transfer info, when it is not set using my dynamic matrix_s",
                    "clip = clip if not core.text.FrameProps(clip, '_Transfer') else core.std.SetFrameProps(clip, _Transfer=5)",
                    "",
                    "# Setting color primaries info, when it is not set",
                    "clip = clip if not core.text.FrameProps(clip, '_Primaries') else core.std.SetFrameProps(clip, _Primaries=5)",
                    "",
                    "# Setting color range to TV (limited) range",
                    "clip = core.std.SetFrameProp(clip=clip, prop='_ColorRange', intval=1)",
                    "",
                    "# Setting frame to progressive",
                    "clip = core.std.SetFrameProp(clip=clip, prop='_FieldBased', intval=0)",
                    "",
                    "# Adjusting color format to RGBS for custom section",
                    "clip = core.resize.Bicubic(clip=clip, format=vs.RGBS, range_in_s='full')",
                    "",
                    "# Apply SCDetect for scene detection",
                    f"clip = core.misc.SCDetect(clip, threshold={scene_change_threshold})",
                    "",
                    f"# Apply RIFE interpolation with calculated multiplier {multi.numerator}/{multi.denominator}",
                    f"clip = vsmlrt.RIFE(clip,",
                    f"    multi=Fraction({multi.numerator}, {multi.denominator}),",  # Using the calculated multi Fraction
                    f"    model=RIFEModel.{model_version},",
                    f"    backend=Backend.TRT(",
                    f"        fp16={rife_params['fp16']},",
                    f"        num_streams={rife_params['num_streams']},",
                    f"        use_cuda_graph={rife_params['use_cuda_graph']},",
                    f"        output_format={rife_params['output_format']},",
                    f"        tf32=True,",
                    f"        heuristic=True",
                    "    ),",
                    f"    ensemble={str(rife_params['ensemble'])},",
                    f"    video_player={str(rife_params['video_player_mode'])}",
                    ")",
                    "",
                    "# Set the correct output FPS using AssumeFPS",
                    f"clip = clip.std.AssumeFPS(fpsnum={target_fps.numerator}, fpsden={target_fps.denominator})",
                    "",
                    "# Remove padding",
                    "clip = clip.std.Crop(right=pw, bottom=ph)",
                ])  , where "rife_params" is just my module to apply custom parameters, the scene detection is added here by SCDetect but without the sc = True , i get stutters and incorrect scene changes, as there is literally no code to use the SCDetect analysis with vsmlrt. How to do this? Thank you.

**Comment by WolframRhodium** [comment]
[Here](https://github.com/AmusementClub/vs-mlrt/blob/2adfbab790eebe51c62c886400b0662570dfe3e9/scripts/vsmlrt.py#L1243-L1250) is the code that handles scene change, which is always enabled.

**Reply by VisionHDR** [reply]
Thanks, so SCDetect is indeed required. I'm still experiencing stutter since I'm not sure how to properly set the multiplier to do fractions from my chosen Framerate. Let's say I need to interpolate from 23.976 to 60 fps, what I should set in the script? f"clip = vsmlrt.RIFE(clip,",
f" multi=Fraction (what goes here then?)
f" model=RIFEModel.{model_version},",
f" backend=Backend.TRT(",
f" fp16={rife_params['fp16']},",
f" num_streams={rife_params['num_streams']},",
f" use_cuda_graph={rife_params['use_cuda_graph']},",
f" output_format={rife_params['output_format']},",
f" tf32=True,",
f" heuristic=True",
" ),",
f" ensemble={str(rife_params['ensemble'])},",
f" video_player={str(rife_params['video_player_mode'])}",
")",
"",
"# Set the correct output FPS using AssumeFPS",
f"clip = clip.std.AssumeFPS(fpsnum={target_fps.numerator}, fpsden={target_fps.denominator})",
"",
"# Remove padding",
"clip = clip.std.Crop(right=pw, bottom=ph)",
])

**Reply by WolframRhodium** [reply]
```python3
from fractions inport Fraction
clip = vsmlrt.RIFE(clip, multi=60 * Fraction(1001, 24000), ......)
```

**Reply by VisionHDR** [reply]
> ```python
> from fractions inport Fraction
> clip = vsmlrt.RIFE(clip, multi=60 * Fraction(1001, 24000), ......)
> ```

That makes it 1000 fps. In super slow motion.

**Reply by WolframRhodium** [reply]
```python3
>>> import vapoursynth as vs
>>>from vapoursynth import core
>>> src = core.std.BlankClip(format=vs.RGBS, fpsnum=24000, fpsden=1001)
>>> import vsmlrt
>>> from fractions import Fraction
>>> flt = vsmlrt.RIFE(src, multi=60*Fraction(1001, 24000))
>>> flt.fps
Fraction(60, 1)
```

**Reply by VisionHDR** [reply]
> ```python
> >>> import vapoursynth as vs
> >>>from vapoursynth import core
> >>> src = core.std.BlankClip(format=vs.RGBS, fpsnum=24000, fpsden=1001)
> >>> import vsmlrt
> >>> from fractions import Fraction
> >>> flt = vsmlrt.RIFE(src, multi=60*Fraction(1001, 24000))
> >>> flt.fps
> Fraction(60, 1)
> ```

Thanks, i managed to get it to: 

 multi=Fraction(60, 1) / Fraction(24000, 1001) , which makes it 60 out of 23.976.

**Reply by VisionHDR** [reply]
> > ```python
> > >>> import vapoursynth as vs
> > >>>from vapoursynth import core
> > >>> src = core.std.BlankClip(format=vs.RGBS, fpsnum=24000, fpsden=1001)
> > >>> import vsmlrt
> > >>> from fractions import Fraction
> > >>> flt = vsmlrt.RIFE(src, multi=60*Fraction(1001, 24000))
> > >>> flt.fps
> > Fraction(60, 1)
> > ```
> 
> Thanks, i managed to get it to:
> 
> multi=Fraction(60, 1) / Fraction(24000, 1001) , which makes it 60 out of 23.976.

After days of intense testing, I think the scene detection code used in vsmlrt is wrong. If the proper method to deal with scene changes is to call SCDetect(clip) before RIFE interpolation this makes scene changes proper, but there are several stuttering and Framerate is never smooth in fast scenes. If I try to use a threshold, like 0 15, smoothness is better but several scene changes are interpolated making the morphing effect. I think the scene detection code (akarin) is lacking proper still frame detection and proper motion detection to deal with both situations. It's currently useless for me the way it works right now.

**Reply by WolframRhodium** [reply]
The akarin part is not responsible for scene detection.

**Reply by VisionHDR** [reply]
> The akarin part is not responsible for scene detection.

Yes, but you said it "deals" with the results from the required SCDetect. Otherwise, how you use it then?

**Reply by WolframRhodium** [reply]
There are other scene detection approaches.

**Reply by VisionHDR** [reply]
> There are other scene detection approaches.

Like SCDetection with super/analyse/smooth? I tried, didn't work. How would you use a proper scene detection? It should set the property for the akarin code in vsmlrt or how would you do? It would be nice to get a working solution, I can pay for it. Thanks.

**Reply by WolframRhodium** [reply]
We don't provide such services. Ultimately you can manually annotate scene changes.
